### PR TITLE
Finish request early: frankenphp_finish_request

### LINKED ISSUE #62: Consider implementing fastcgi_finish_request()?
[`fastcgi_finish_request`](https://www.php.net/manual/en/function.fastcgi-finish-request.php) is often used to perform operations after a request is completed (such as sending an email). Should FrankenPHP implement this and/or create a similar function? There are pro/cons to implementing it with the same name (it will "just work" with existing code), and a con would be that if there are any differences in behavior, then things may not work as expected or code may use the existence of that function as a sentinel that it is running in a fastcgi context and do unexpected things.

I'd be happy to tackle implementing this, whatever is decided.

### PR DESCRIPTION
**Pull request by withinboredom**
This adds a new function: `frankenphp_finish_request` and an alias for it: `fastcgi_finish_request`, which follows in the footsteps of [other SAPIs](https://github.com/php/php-src/blob/master/sapi/litespeed/lsapi_main_arginfo.h#L35).

closes #62

### CODE REVIEW

File: frankenphp_arginfo.h [hunk 1]
```diff
@@ -9,13 +9,18 @@ ZEND_BEGIN_ARG_WITH_RETURN_TYPE_INFO_EX(arginfo_headers_send, 0, 0, IS_LONG, 0)
...
 ZEND_END_ARG_INFO()
 
+ZEND_BEGIN_ARG_WITH_RETURN_TYPE_INFO_EX(arginfo_frankenphp_finish_request, 0, 0, _IS_BOOL, 0)
+ZEND_END_ARG_INFO()
+
 
 ZEND_FUNCTION(frankenphp_handle_request);
 ZEND_FUNCTION(headers_send);
-
+ZEND_FUNCTION(frankenphp_finish_request);
```
Comment (withinboredom) [inline frankenphp_arginfo.h:18]: I couldn't quite figure out how to generate this file automatically, so this is currently hand-crafted atm.

File: frankenphp.stub.php [hunk 2]
```diff
@@ -5,3 +5,5 @@
 function frankenphp_handle_request(callable $callback): bool {}
 
 function headers_send(int $status = 200): int {}
+
+function frankenphp_finish_request(): bool {}
```
Comment (dunglas) [inline frankenphp.stub.php:9]:
```suggestion
function frankenphp_finish_request(): bool {}
```

File: testdata/finish-request.php [hunk 3]
```diff
@@ -0,0 +1,11 @@
...
+
+echo "<pre>I am here\n";
+
+var_dump(frankenphp_finish_request());
+
+echo "\n";
+
+var_dump(fastcgi_finish_request());
+
+echo "\nI am here too";
```
Comment (dunglas) [inline testdata/finish-request.php:11]:
```suggestion
echo "\nI am here too";
```

File: frankenphp.c [hunk 4]
```diff
@@ -55,6 +55,7 @@ typedef struct frankenphp_server_context {
 	uintptr_t current_request;
 	uintptr_t main_request;				/* Only available during worker initialization */
 	char *cookie_data;
+	bool finished;
```
Comment (withinboredom) [inline frankenphp.c:58]: We need some way to track whether the request is "finished" (i.e., no more output and output flushed) or not.

File: frankenphp.c [hunk 5]
```diff
@@ -165,6 +170,27 @@ static int frankenphp_worker_request_startup() {
...
+    		RETURN_THROWS();
+    }
+
+    frankenphp_server_context *ctx = SG(server_context);
+
+    if(!ctx->finished) {
+    	php_output_end_all();
+    	php_header();
+
+    	go_frankenphp_worker_handle_request_end(ctx->current_request, false);
```
Comment (withinboredom) [inline frankenphp.c:184]: I'm not 100% sure about this function, but it appears to be what is required for both worker/non-worker mode. @dunglas, I know a bit about Go to be dangerous, but I'm not an expert. I'm going to have to rely on you if this is wrong.

File: frankenphp.go [hunk 6]
```diff
@@ -359,6 +362,13 @@ func go_fetch_request() C.uintptr_t {
 	return C.uintptr_t(cgo.NewHandle(r))
 }
 
+func maybeCloseContext(fc *FrankenPHPContext) {
+	if !fc.closed {
+		close(fc.done)
+		fc.closed = true
+	}
+}
+
```
Comment (withinboredom) [inline frankenphp.go:371]: It appears we close this channel in a couple of places in the code, which isn't ideal when we want to make sure we close it exactly once.

File: frankenphp_test.go [hunk 7]
```diff
@@ -100,6 +100,20 @@ func testHelloWorld(t *testing.T, opts *testOptions) {
 	}, opts)
 }
 
+func TestFinishRequest_module(t *testing.T) { testFinishRequest(t, nil) }
+//func TestFinishRequest_worker(t *testing.T) { testFinishRequest(t, &testOptions{workerScript: "finish-request.php"}) }
```
Comment (withinboredom) [inline frankenphp_test.go:104]: This goes into an infinite loop, and I'm not sure why yet. It appears to work fine "in real life" but there must be something wrong. I'll dig into it.

File: frankenphp.go (hunk 6 above)
Comment (dunglas) [reply frankenphp.go:371]: You can probably use https://pkg.go.dev/sync#Once for this use case.

**Comment by chalasr** [comment]
Is the `fastcgi_finish_request`  alias really needed? What does it bring?

**Comment by withinboredom** [comment]
> Is the fastcgi_finish_request alias really needed? What does it bring?

Mostly it brings compatibility with other scripts/libraries without needing to make specific changes just to support frankenphp.

**Comment by chalasr** [comment]
Other SAPIs don't have such alias though (or even [removed it](https://github.com/php/php-src/commit/ccf051c317e606c2a4d9099c6e79a5c42bfdb298) for some reason)

**Comment by dunglas** [comment]
NGINX Unit implements this function (with no alias): https://github.com/nginx/unit/blob/master/src/nxt_php_sapi.c#L216

Having it will improve the performance of the hundreds of applications out there relying on this feature.

**Comment by dunglas** [comment]
And according to https://github.com/nginx/unit/issues?q=is%3Aissue+fastcgi_finish_request (no bug report since this feature has been implemented), it's quite safe to do as long as this function behaves exactly as the original one (which should be the case here).

Also, if someone really want to exclude FrankenPHP, it's quite easy to do using feature/SAPI detection.

**Comment by chalasr** [comment]
Makes sense. Thank you guys for the explanations

**Review by dunglas — APPROVED** [review]

**Comment by withinboredom** [comment]
🎉

**Pull request merged**